oci/layout/oci_delete.go

```diff
@@ -0,0 +1,244 @@
+	"os"
+
+	"github.com/containers/image/v5/internal/set"
+	"github.com/containers/image/v5/types"
+	digest "github.com/opencontainers/go-digest"
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+)
+
+// DeleteImage deletes the named image from the directory, if supported.
+func (ref ociReference) DeleteImage(ctx context.Context, sys *types.SystemContext) error {
+	sharedBlobsDir := ""
+	if sys != nil && sys.OCISharedBlobDirPath != "" {
+		sharedBlobsDir = sys.OCISharedBlobDirPath
+	}
+
+	desciptor, err := ref.getManifestDescriptor()
+	if err != nil {
+		return err
+	}
+
+	switch desciptor.MediaType {
+	case imgspecv1.MediaTypeImageManifest:
+		return ref.deleteSingleImage(&desciptor, sharedBlobsDir)
+	case imgspecv1.MediaTypeImageIndex:
+		return ref.deleteImageIndex(&desciptor, sharedBlobsDir)
+	default:
+		return fmt.Errorf("unsupported mediaType in index")
+	}
+}
+
+func (ref ociReference) deleteSingleImage(desciptor *imgspecv1.Descriptor, sharedBlobsDir string) error {
+	manifest, err := ref.getManifest(desciptor, sharedBlobsDir)
+	if err != nil {
+		return err
+	}
+	blobsUsedInManifest := ref.getBlobsUsedInManifest(manifest)
+	blobsUsedInManifest[desciptor.Digest]++ // Add the current manifest to the list of blobs used by this reference
+
+	blobsToDelete, err := ref.getBlobsToDelete(blobsUsedInManifest, sharedBlobsDir)
+	if err != nil {
+		return err
+	}
+
+	err = ref.deleteBlobs(blobsToDelete)
+	if err != nil {
+		return err
+	}
+
+	return ref.deleteReferenceFromIndex()
+}
+
+func (ref ociReference) deleteImageIndex(desciptor *imgspecv1.Descriptor, sharedBlobsDir string) error {
+	blobPath, err := ref.blobPath(desciptor.Digest, sharedBlobsDir)
+	if err != nil {
+		return err
+	}
+	index, err := parseIndex(blobPath)
+	if err != nil {
+		return err
+	}
+
+	blobsUsedInImageRefIndex, err := ref.getBlobsUsedInIndex(index, sharedBlobsDir)
+	if err != nil {
+		return err
+	}
+	blobsUsedInImageRefIndex[desciptor.Digest]++ // Add the nested index in the list of blobs used by this reference
+
+	blobsToDelete, err := ref.getBlobsToDelete(blobsUsedInImageRefIndex, sharedBlobsDir)
+	if err != nil {
+		return err
+	}
+
+	err = ref.deleteBlobs(blobsToDelete)
+	if err != nil {
+		return err
+	}
+
+	return ref.deleteReferenceFromIndex()
+}
+
+// Returns a map of digest with the usage count, so a blob that is referenced three times will have 3 in the map
+func (ref ociReference) getBlobsUsedInIndex(index *imgspecv1.Index, sharedBlobsDir string) (map[digest.Digest]int, error) {
+	blobsUsedInIndex := make(map[digest.Digest]int)
+
+	for _, desciptor := range index.Manifests {
+		blobsUsedInIndex[desciptor.Digest]++
+		switch desciptor.MediaType {
+		case imgspecv1.MediaTypeImageManifest:
+			manifest, err := ref.getManifest(&desciptor, sharedBlobsDir)
+			if err != nil {
+				return nil, err
+			}
+			for digest, count := range ref.getBlobsUsedInManifest(manifest) {
+				blobsUsedInIndex[digest] += count
+			}
```

Comment:
Indeed; I made the change in a separate commit. However, I never liked functions that modify the arguments they are given, as a matter of taste/readability (I mostly worked with languages where this was discouraged).

If this is habitual in Go and in this codebase, I'll roll with it :-)

Reply:
This is non-blocking, I really doubt the performance difference will matter.

---

(I think, in principle, it would be idiomatic to have a
```
class blobCounter {
   countsSoFar: mutable map[Digest]int
}
```
and some `usedBlob`/`addBlobsUsedInIndex` methods on that `blobCounter` class, in almost all object-oriented  (not pure-functional). languages

And then I think that that a single-purpose single-field class is just a pure overhead, to readability most of all, so passing that map around without the class wrapper is just simpler.

But it _does_ require careful naming and/or documentation to not be surprising. I’m definitely not insisting on that approach.